README.md
# KaradaColor

<!-- badges: start -->
[![cran version](https://www.r-pkg.org/badges/version/KaradaColor?color=orange)](https://cran.r-project.org/package=KaradaColor)
[![rstudio mirror per-month downloads](http://cranlogs.r-pkg.org/badges/last-month/KaradaColor?color=yellow)](https://cran.r-project.org/package=KaradaColor)
[![rstudio mirror total downloads](http://cranlogs.r-pkg.org/badges/grand-total/KaradaColor?color=blue)](https://cran.r-project.org/package=KaradaColor)
<!-- badges: end -->

Color palette of "[KARADA-GOOD](https://www.karada-good.net/)". The color palette consists of five colors selected from photographs of Japanese landscapes and foods. Color palette will be added sequentially.

Color palettes to be added are introduced on [Instagrame](https://www.instagram.com/karadagood/) from time to time.


## Installation
Get the released version from CRAN:
``` r
install.packages("KaradaColor")
```
Or the development version from github:
``` r
# install.packages("devtools")
devtools::install_github("KaradaGood/KaradaColor")
```

## Example

``` r
library(KaradaColor)
#Display all palettes in console
kg_show_console()
```

<img src="man/figures/Show_KG_Palette.png" width="50%"/>

``` r
#Plot the palette
kg_plot_color("Kyoto_City", n = 5, showcode = TRUE)
```

<img src="man/figures/Plot_KG_Palette.png" width="50%"/>

``` r
#Create complementary color: kg_create_cc() command
kg_plot_color(color = kg_create_cc())
```

<img src="man/figures/kg_create_cc.png" width="50%"/>

``` r
#Create triad color: kg_create_tc() command
kg_plot_color(color = kg_create_tc())
```

<img src="man/figures/kg_create_tc.png" width="50%"/>

``` r
#Monochromatic Color: kg_create_mc() command
kg_plot_color(color = kg_create_mc(basecols = "#d0af4c"))
```
<img src="man/figures/kg_create_mc.png" width="50%"/>


## ggplot2
The package contains colour scale functions for `scale_color_kg()` and `scale_fill_kg()`. 
Missing values are initially displayed in red.

``` r
#Continuous scale exsample
if(!require("tidyverse", quietly = TRUE)){
  install.packages("tidyverse");require("tidyverse")
}
set.seed(1)
x <- LETTERS[1:20]
y <- paste0("var", seq(1,20))
data <- expand.grid(X=x, Y=y)
data$Z <- sample(c(NA, rnorm(15)), 20, replace = TRUE)
library("tidyverse")
ggplot(data, aes(X, Y, fill= Z)) +
  geom_tile() +
  scale_fill_kg(discrete = FALSE, name = "Hanamushiro",
                alpha = 1, na.value = "red")
```

<img src="man/figures/scale_fill_KG.png" width="50%"/>

``` r
#Discrete scale exsample
library("ggplot2")
ggplot(data = diamonds, aes(x = cut, y = price,
                            color = cut, fill = cut)) +
  geom_boxplot() +
  scale_color_kg(name = "Hanamushiro", alpha = 0.3) +
  scale_fill_kg(name = "Hokkaido_Sky", alpha = 1) +
  theme_dark()
```

<img src="man/figures/scale_color_KG.png" width="50%"/>

## aRtsy and other packages
This can be done with the kg_get_color() command.
``` r
library("aRtsy")
set.seed(1234)
canvas_strokes(colors = kg_get_color(name = "Otaru_Unga"),
               neighbors = 1, p = 0.02, iterations = 1,
               resolution = 350)
```

<img src="man/figures/get_KG_color.png" width="50%"/>


NEWS.md
# KaradaColor 0.1.6
Add:kg_create_mc() command. This function create monochromatic color.

# KaradaColor 0.1.5
Release to CRAN
Change:Rename some commands to kg_xxx_xxx before allowing CRAN release.

# KaradaColor 0.1.4
Change:kg_plot_color() command. Support for color vector data.
Add:convert_colTodf() command. This function convert color data to data.frame class.
Add:image_add_textAcode() command. This function Add to image() command text and color code.

# KaradaColor 0.1.3

Add:kg_create_cc() command. This function create complementary color.
Change:show_KG_palette() command. Support for data.frame class.

# KaradaColor 0.1.2

Initial release at git.

# KaradaColor 0.1.1

A memorable line. \* Added a `NEWS.md` file to track changes to the package.


R/Use_KG_Palette.R
#' Get color palette data
#'
#' \code{get_KG_color} This function creates a vector of \code{n} color palette data for the KaradaColor Package.
#' @importFrom grDevices colorRamp rgb
#' @param n Number of colors to create.
#' @param name color palette name. (default: \code{"Sapporo_Sta"})
#' @param alpha	The alpha transparency. Numbers range from 0-1. see argument alpha in \code{\link[grDevices]{rgb}}.
#' @return Vector containing a hex color code representation for the chosen palette.
#' @author Masaki Fujishima \email{info@karada-good.net}, \href{https://github.com/KaradaGood/KaradaColor}{Github}, \href{https://www.karada-good.net/analyticsr/r-770/}{Personal Blog}
#' @export get_KG_color
#' @examples
#' library("scales")
#' show_col(get_KG_color(n = 20, name = "Totan", alpha = 0.5))
#'
#' library("aRtsy")
#' set.seed(1234)
#' canvas_strokes(colors = get_KG_color(name = "Otaru_Unga"),
#'                neighbors = 1, p = 0.02, iterations = 1,
#'                resolution = 350)
get_KG_color <- function(n = 5, name = "Sapporo_Sta", alpha = 1){

  pal_col <- df_KG_palettes[[name]]
  KG_cols <- grDevices::colorRamp(pal_col, space = "Lab", interpolate = "spline")
  cols <- KG_cols(seq(0, 1, length.out = n)) / 255
  return(grDevices::rgb(cols[, 1], cols[, 2], cols[, 3], alpha = alpha))

  }

#' @rdname get_KG_color
#' @importFrom grDevices colorRamp rgb
#' @export
get_KG_palette <- function(name = "Sapporo_Sta", alpha = 1){

  function(n) {

    return(get_KG_color(n, name, alpha))

    }
}

#' Use KaradaColor scales on ggplot2
#'
#' {scale_color_KG} Use KaradaColor color palette on ggplot2 colors.
#' @importFrom ggplot2 discrete_scale scale_color_gradientn
#' @param name color palette name. (default: \code{"Sapporo_Sta"})
#' @param alpha	The alpha transparency. Numbers range from 0-1. see argument alpha in \code{\link[grDevices]{rgb}}.
#' @param na.value Color of missing value.
#' @param discrete Generate a discrete palette. (\code{FALSE} - generate continuous palette)
#' @return Create a user defined color palette for ggplot2.
#' @author Masaki Fujishima \email{info@karada-good.net}, \href{https://github.com/KaradaGood/KaradaColor}{Github}, \href{https://www.karada-good.net/analyticsr/r-770/}{Personal Blog}
#' @export scale_color_KG
#' @examples
#' library("ggplot2")
#' ggplot(data = diamonds, aes(x = cut, y = price,
#'                             color = cut, fill = cut)) +
#'   geom_boxplot() +
#'   scale_color_KG(name = "Hanamushiro", alpha = 0.2) +
#'   scale_fill_KG(name = "Hokkaido_Sky", alpha = 0.2) +
#'   theme_dark()
scale_color_KG <- function(name = "Sapporo_Sta", alpha = 1,
                           na.value = "red", discrete = TRUE) {

  if(discrete){

    ggplot2::discrete_scale(aesthetics = "color",
                            scale_name = name,
                            palette = get_KG_palette(name, alpha),
                            na.value = na.value)

  }else{

    ggplot2::scale_color_gradientn(colours = get_KG_color(n = 255, name, alpha),
                                   na.value = na.value)

  }

 }

#' @rdname scale_color_KG
#' @importFrom ggplot2 discrete_scale
#' @export
scale_colour_KG <- scale_color_KG

#' Use KaradaColor fills on ggplot2
#'
#' \code{scale_fill_KG} Use KaradaColor color palette on ggplot2 fills.
#' @importFrom ggplot2 discrete_scale scale_fill_gradientn
#' @param name color palette name. (default: \code{"Sapporo_Sta"})
#' @param alpha	The alpha transparency. Numbers range from 0-1. see argument alpha in \code{\link[grDevices]{rgb}}.
#' @param na.value Color of missing value.(default: "red")
#' @param discrete Generate a discrete palette. (\code{FALSE} - generate continuous palette)
#' @return Create a user defined color palette for ggplot2.
#' @author Masaki Fujishima \email{info@karada-good.net}, \href{https://github.com/KaradaGood/KaradaColor}{Github}, \href{https://www.karada-good.net/analyticsr/r-770/}{Personal Blog}
#' @export scale_fill_KG
#' @examples
#' set.seed(1)
#' x <- LETTERS[1:20]
#' y <- paste0("var", seq(1,20))
#' data <- expand.grid(X=x, Y=y)
#' data$Z <- sample(c(NA, rnorm(15)), 20, replace = TRUE)
#' library("ggplot2")
#' ggplot(data, aes(X, Y, fill= Z)) +
#'   geom_tile() +
#'   scale_fill_KG(discrete = FALSE, name = "Hanamushiro",
#'                 alpha = 0.4, na.value = "red")
scale_fill_KG <- function(name = "Sapporo_Sta", alpha = 1,
                          na.value = "red", discrete = TRUE) {

  if(discrete){

    ggplot2::discrete_scale(aesthetics = "fill",
                            scale_name = name,
                            palette = get_KG_palette(name, alpha),
                            na.value = na.value)

  }else{

    ggplot2::scale_fill_gradientn(colours = get_KG_color(n = 255, name, alpha),
                                  na.value = na.value)

  }

}
